Refactor low-level UTF-16 decoding and move Borrow(Mut) to libcore.

SimonSapin opened the pull request:
* Rename `Utf16Items` to `Utf16Decoder`. "Items" is meaningless.
* Generalize it to any `u16` iterator, not just `[u16].iter()`
* Make it yield `Result` instead of a custom `Utf16Item` enum that was isomorphic to `Result`. This enable using the `FromIterator for Result` impl.
* Replace `Utf16Item::to_char_lossy` with a `Utf16Decoder::lossy` iterator adaptor.

This is a [breaking change], but only for users of the unstable `rustc_unicode` crate.

I’d like this functionality to be stabilized and re-exported in `std` eventually, as the "low-level equivalent" of `String::from_utf16` and `String::from_utf16_lossy` like #27784 is the low-level equivalent of #27714.

CC @aturon, @alexcrichton

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

SimonSapin commented on src/libserialize/json.rs:
```diff
@@ -1713,11 +1711,13 @@ impl<T: Iterator<Item=char>> Parser<T> {
+                                return self.error(LoneLeadingSurrogateInHexEscape)
                             }
+                            let c = (((n1 - 0xD800) as u32) << 10 |
+                                     (n2 - 0xDC00) as u32) + 0x1_0000;
```
Since we’re only dealing with a single pair of `u16`s here, I felt this was simpler than creating a one-shot iterator. But that’s still possible.

eefriedman commented on src/librustc_unicode/u_str.rs:
```diff
@@ -413,67 +412,82 @@ pub fn is_utf16(v: &[u16]) -> bool {
+    /// use rustc_unicode::str::Utf16Decoder;
+    ///
+    /// fn main() {
+    ///     // 𝄞mus<invalid>ic<invalid>
```
My computer (Windows 8.1) doesn't have a font containing U+1D11E; maybe pick something a bit less exotic?

SimonSapin replied:
(I’m just moving this comment, it was previously on the `utf16_items` function: http://doc.rust-lang.org/rustc_unicode/str/fn.utf16_items.html)

SimonSapin commented:
I’d like re-exporting `Utf16Decoder` and `Utf16LossyDecoder` in `std` while I’m at it (still unstable for now), but I don’t know what module they would go into. Maybe `std::char`, since they (sort of) return `char` iterators?

bors commented:
:umbrella: The latest upstream changes (presumably #27684) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Rebased and added commits to move the iterators to `rustc_unicode::char` (reexported/exposed in `std::char`) with an unstable `utf16_decoder` feature name (#27830), and add issue annotations per #27818.

bluss commented:
Should be `[breaking-change]` with hyphen in the message.

SimonSapin commented:
@bluss fixed

alexcrichton commented:
Thanks for the PR @SimonSapin! I like the reduction in API surface area here, and I think that this can actually go even farther and remove the lossy decoder (as it's just a simple `.map(...)` iterator anyway). I'm not 100% certain that this is an API we'd want to expose as stable at some point (we have little precedent for iterator constructors today), but I have no outstanding issue with the API as-is.

Could you leave behind the old support, but mark it all as `#[deprecated]`? (to help ease current users on nightly).

SimonSapin commented:
* Added deprecated wrappers for the old API
* Removed the lossy decoder
* Added a `std::char::REPLACEMENT_CHARACTER` const instead, and documented how to do lossy decoding with it.
* Moved `Utf16Decoder::new` to `Iterator::decode_utf16`, which seems more idiomatic for iterator adaptors. That moves it to libcore and (indirectly) into the prelude, but then that is already the case for `char::encode_utf16`. It’s also the first `Iterator` method that only works with one concrete `Iterator::Item` type.

I’m not so sure about that last change. What do you think?

pnkfelix commented:
@SimonSapin sorry, I was on PTO for much of the time this PR was open.

I'll skim over it but I'm not certain I'm the best person to review.  (I assume the Travis failure is due to a timeout, though I was under a rough impression that timeouts were not supposed to end up being marked as fatal checks on the PRs...)

SimonSapin commented:
No worries @pnkfelix, @alexcrichton took over.

After a chat with @bluss on IRC I undid the move to `Iterator::decode_utf16` method (although method chaining is nice, this is not the right place for it) and moved to a `std::char::decode_utf16` function instead. I also generalized `u16` to `Borrow<u16>` (moving `Borrow` to libcore) and (in the constructor function) `Iterator` to `IntoIterator`. This allows using a `&[u16]` slice directly, like before.

bluss commented:
Nice. I think orrow should move to libcore anyway, since it is becoming a commonly used trait, not just for collections. I think the docs are clear enough for Borrow, to make its important guarantees clear.

bluss commented:
We also found out free functions make sense to construct iterators because the focus is not at all on the struct carrying the iteration, that's almost an implementation detail. And it's consistent with libstd's precedents (std::ascii::escape_default)

SimonSapin commented:
(Before moving just `Borrow` and `BorroMut` I tried moving the entire `collections::borrow` module, but couldn’t because of coherence rules.)

alexcrichton commented:
I think it's fine to move the `Borrow` and `BorrowMut` traits into libcore, but I feel like it's a bit too much magic to rely on it in iterators currently. The only precedent we have for being generic over `Borrow` is just in collections, and starting to extend that to iterators isn't quite as idiomatic. For example the `cloned` iterator adapter doesn't strive to use `Borrow` even though it's technically all it needs.

alexcrichton commented:
I'm also ok with a free function for now to construct this iterator instead of `Iterator::new`, but I don't quite feel that the `char` module is the best place for a stable location for this eventually. Unfortunately, though, I don't think we have a great place for it to be located, so it at least works for an unstable location for now.

SimonSapin commented:
`Borrow<u16>` is used here as a way to express “`u16` or `&u16`”. The former is the most general case, `&u16` doesn’t work if you’re generating `u16`’s incrementally without allocating buffers. The latter, when combined with `IntoIterator`, conveniently allows passing a `&[u16]` slice directly.

I don’t care so much about generalizing `&u16` to other kinds of references.

If you feel this is too much magic I can reverse to just `u16`, but `decode_utf16(slice)` would have to be more verbose: `decode_utf16(slice.iter().cloned())`.

alexcrichton commented:
I definitely agree that `decode_utf16(slice)` is more ergonomic than the alternative, but this is also a pretty niche function that's very rarely called, so I don't think it's necessarily important to have the best ergonomics right off the bat (especially when it's not an established convention in the standard library).

SimonSapin commented:
Alright, I’ll revert the use of `Borrow`. Should I keep the move of `Borrow` to libcore? It becomes unrelated to this PR.

alexcrichton commented:
I'm fine either way on that, it'll probably happen eventually regardless.

bors commented:
:umbrella: The latest upstream changes (presumably #27871) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Rebased.

aturon commented:
LGTM!

alexcrichton commented:
We talked about this at the libs meeting today and decided to move forward, thanks @SimonSapin!

@bors: r+ 6174b8d726ed5764694e5404329d8b5e66517ed5

bors commented:
:hourglass: Testing commit 6174b8d with merge fd302a9...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6214), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6228), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6168), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6194), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6172), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6246), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6240), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6204), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1188), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1187), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1186), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1190), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/469), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1168)
